### PR TITLE
Emit more information for summarizer observability

### PR DESCRIPTION
**Pull request by kian-thompson**
By providing a way to programmatically observing summarization failures we can make progress in mitigating/reducing summarizer related issues. Customer OCE's will be able to catch a summarization bugs on their own and decide what to do with their end-users (either alert them via UI, alert themselves via Telemetry, etc.). Customers can also do their own summarization testing, and thus catch some of these bugs before they ship.

Included in this change are new events being emitted off the `ContainerRuntime` class and `IContainerRuntime` interface. They include:
- `"summarizeAllAttemptsFailed"`: All attempts to summarize have failed.
- `"summarizerStop"`: The summarizer has stopped, providing information on why and potential error.
- `"summarizerStart"`: The summarizer has started.
- `"summarizerStartupFailed"`: The summarizer has failed to startup, usually meaning it wasn't able to catchup. Includes information on why startup failed.

[AB#24393](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/24393)

### CODE REVIEW

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM

File: packages/runtime/container-runtime/api-report/container-runtime.legacy.alpha.api.md [hunk 1]
```diff
@@ -545,6 +545,10 @@ export interface ISummarizeEventProps {
     // (undocumented)
     error?: any;
     // (undocumented)
+    failureMessage?: string;
+    // (undocumented)
+    isLastSummary?: string;
```
Comment (jatgarg) [inline packages/runtime/container-runtime/api-report/container-runtime.legacy.alpha.api.md:550]: nit: Can you add a little comment for this?

**Comment by jatgarg** [comment]
@jason-ha Can this PR go in before breaking change release as we are adding to legacy?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -2003,9 +2003,19 @@ export class ContainerRuntime
...
+				// Forward events from SummaryManager
+				[
+					"summarize",
+					"summarizeAllAttemptsFailed",
+					"summarizerStop",
+					"summarizerStart",
+					"summarizerStartupFailed",
+				].forEach((eventName) => {
+					this.summaryManager?.on(eventName, (...args: any[]) => {
+						this.emit(eventName, ...args);
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2015]: you emit these, but they are not exposed in the container runtime typing at all. I also wonder what the expected usage of these events is, and if each really needs it own event, or if we could have a single event, but a discriminated union of property objects. if the goal is logging, i think a combined even is easier to managed.

consider adding a scenario test under packages\test\local-server-tests to illustrate usage.

**Comment by jason-ha** [comment]
> @jason-ha Can this PR go in before breaking change release as we are adding to legacy?

It is okay to extend the functionality of legacy (and public) interfaces that are not externally implemented. These should be marked with `@sealed` tag to indicate that only implementations should come from FF. If there are any places where the modified interface is input to a FF API, then some more caution may be needed when the interface is not already marked sealed. Maybe someone decided they should inject a proxy or such and that would lead to trouble.

Always good to go survey legacy users code to see where there may be impact.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2015]: This is a good point. I'll move the typing to be off the `IContainerRuntime` interface instead.
These events are to help consumers (i.e. Loop) better observe when the summarizer is failing.

File: .changeset/dirty-crabs-try.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+"@fluidframework/container-runtime": minor
+---
+---
+"section": deprecation
+---
+
+Deprecated SummarizerStopReason, ISummarizeEventProps, and ISummarizerEvents
+
+`SummarizerStopReason`, `ISummarizeEventProps`, and `ISummarizerEvents` have all been deprecated from the `"@fluidframework/container-runtime"` package. Please migrate all usage of these APIs to their respective copies in the `"@fluidframework/container-runtime-definitions"` package.
```
Comment (jzaffiro) [inline .changeset/dirty-crabs-try.md:10]:
```suggestion
`SummarizerStopReason`, `ISummarizeEventProps`, and `ISummarizerEvents` have all been deprecated from the `"@fluidframework/container-runtime"` package. Please migrate all uses of these APIs to their respective copies in the `"@fluidframework/container-runtime-definitions"` package.
```
nit

**Review by jzaffiro — APPROVED** [review]
Changeset looks good pending one minor comment!

**Review by jatgarg — APPROVED** [review]

**Pull request merged**